examples/step-79/step-79.cc

```diff
@@ -0,0 +1,2444 @@
+
+#include <deal.II/numerics/vector_tools.h>
+#include <deal.II/numerics/matrix_tools.h>
+#include <deal.II/numerics/data_out.h>
+
+
+
+#include <iostream>
+#include <fstream>
+#include <algorithm>
+
+// Above are fairly common files to include. These also include the
+// one for the sparse direct class SparseDirectUMFPACK. This is not
+// the most efficient way to solve large linear problems, but it will
+// do for now.
+//
+// As usual, we put everything into a common namespace. We then start
+// by declaring a number of symbolic names for constants that will be
+// used throughout this tutorial. Specifically, we have a *lot* of
+// variables in this program (of course the density and the displacement,
+// but also the unfiltered density and quite a number of Lagrange multipliers).
+// It is easy to forget which of these variables is at which position in
+// the solution vector, and trying to use numbers for these vector
+// components is a prescription for bugs. Rather, we define static
+// variables that can be used in all of these places and that have to
+// be initialized only once. In practice, this will lead to some
+// lengthy expressions, but they are more readable and less likely to
+// be wrong.
+//
+// A similar issue arises with the ordering of blocks in the system
+// matrix and in vectors. The matrices have $9\times 9$ blocks, and
+// it's difficult to remember which is which. It is far easier to just
+// use symbolic names for those as well.
+//
+// Finally, while we're at it, we introduce symbolic names also for
+// the boundary indicators we will use, in the same spirit as was done
+// in step-19.
+//
+// In all of these cases, we declare these variables as members in a
+// namespace. In the case of the solution components, the concrete
+// values of these variables depend on the space dimension, so we use
+// [template
+// variables](https://en.cppreference.com/w/cpp/language/variable_template)
+// to make the value of the variable depend on a template argument in
+// the same way as we often use template functions.
+namespace SAND
+{
+  using namespace dealii;
+
+  namespace SolutionComponents
+  {
+    template <int dim>
+    constexpr unsigned int density = 0;
+    template <int dim>
+    constexpr unsigned int displacement = 1;
+    template <int dim>
+    constexpr unsigned int unfiltered_density = 1 + dim;
+    template <int dim>
+    constexpr unsigned int displacement_multiplier = 2 + dim;
+    template <int dim>
+    constexpr unsigned int unfiltered_density_multiplier = 2 + 2 * dim;
+    template <int dim>
+    constexpr unsigned int density_lower_slack = 3 + 2 * dim;
+    template <int dim>
+    constexpr unsigned int density_lower_slack_multiplier = 4 + 2 * dim;
+    template <int dim>
+    constexpr unsigned int density_upper_slack = 5 + 2 * dim;
+    template <int dim>
+    constexpr unsigned int density_upper_slack_multiplier = 6 + 2 * dim;
+  } // namespace SolutionComponents
+
+  namespace SolutionBlocks
```

Comment:
Two suggestions (just putting this out there): 
- For `SolutionComponents`, you could make each of these reference the one above so there's less that's hardcoded. All that each entry needs to know is the entry before it and how many components that entry had. That also self-documents that there's a hierarchy that's being constructed.  It you changed the namespace to a struct, then you could also have the template in a single place:
```c++
  template <int dim>
  struct SolutionComponents
  {
    constexpr unsigned int density = 0;
    constexpr unsigned int displacement = density + 1;
    constexpr unsigned int unfiltered_density = displacement + dim;
    constexpr unsigned int displacement_multiplier = unfiltered_density + 1;
    ...
  } // namespace SolutionComponents
```
- As for the `SolutionBlocks`, those could just be an enum with `density=0` and the rest unnumbered. Then you wouldn't need to explicitly write all of that out.

Reply:
And there I was so proud of the template variables in a namespace...

Reply:
The nice thing about the namespace is that, because I typically use a lot of these all at once, I don't have to have SolutionBlocks:: before every one, which might make some of these already hard to read parts even a bit more cumbersome to read ([Like here that Tyler mentioned](https://github.com/dealii/dealii/pull/12034#discussion_r630369908)).